### PR TITLE
feat: add integ test cases for key rotation

### PR DESCRIPTION
**Pull request by sailorlqh**
image

Keys will be destroyed after tests are finished.

### CODE REVIEW

File: test/integ-new/main_new_test.go [hunk 1]
```diff
@@ -212,47 +222,110 @@ func TestUIServiceHealthCheck(t *testing.T) {
...
-		tc := tc
-		t.Run(tc.name, func(t *testing.T) {
-			t.Parallel()
+// Subtests must be run in sequence, and they have waits in between.
+// Therefore, they cannot be parallelized, and aren't a good fit for table testing.
+//
+//nolint:paralleltest
+func TestCertRotatorKeyRotation(t *testing.T) {
+	t.Parallel()
+	testutil.SkipIfNotIntegration(t)
```
Comment (sqin2019) [inline test/integ-new/main_new_test.go:253]: This is not required, as we did the check in TestMain

File: .github/workflows/ci-new.yml [hunk 2]
```diff
@@ -258,7 +270,9 @@ jobs:
       #
       # See: https://cloud.google.com/kms/docs/faq#cannot_delete.
       - name: 'Clean up kms key'
```
Comment (sqin2019) [inline .github/workflows/ci-new.yml:272]: Since we cleaned up in the test, should we remove?

File: test/integ-new/main_new_test.go [hunk 3]
```diff
@@ -297,3 +370,108 @@ func testSendHTTPReq(ctx context.Context, tb testing.TB, uri, token string) *htt
 
 	return resp
 }
+
+func testCallKeyRotationEndpoint(ctx context.Context, tb testing.TB) {
```
Comment (sqin2019) [inline test/integ-new/main_new_test.go:374]: Maybe we can combine testSendHTTPReq, we don't need two similar helper func.

File: test/integ-new/main_new_test.go [hunk 4]
```diff
@@ -212,47 +222,110 @@ func TestUIServiceHealthCheck(t *testing.T) {
...
+		testValidateKeyVersionState(ctx, t, kmsClient, keyResouceName, 2,
+			map[int]kmspb.CryptoKeyVersion_CryptoKeyVersionState{
+				1: kmspb.CryptoKeyVersion_DESTROY_SCHEDULED,
+				2: kmspb.CryptoKeyVersion_ENABLED,
+				3: kmspb.CryptoKeyVersion_ENABLED,
+			})
+	})
+
+	t.Cleanup(func() {
+		testCleanUpKey(ctx, t, kmsClient, keyResouceName)
```
Comment (sqin2019) [inline test/integ-new/main_new_test.go:324]: If we cleanup here, would it affect api and public key tests?

File: .github/workflows/ci-new.yml [hunk 5]
```diff
@@ -37,7 +37,14 @@ env:
...
   KMS_KEYRING_ID: 'jvs-keyring-0666'
-  TAG_ID: 'ci-${{ github.run_id }}-${{ github.run_number }}'
+  KMS_KEY_NAME: 'ci-${{ github.run_id }}-${{ github.run_attempt }}'
+  TAG_ID: 'ci-${{ github.run_id }}-${{ github.run_attempt }}'
+
+  INTEG_TEST_JVS_ROTATION_KEY_TTL: '7s'
+  INTEG_TEST_JVS_ROTATION_GRACE_PERIOD: '2s'
+  INTEG_TEST_JVS_ROTATION_PROPAGATION_DELAY: '1s'
+  INTEG_TEST_JVS_ROTATION_DISABLED_PERIOD: '1s'
+  INTEG_TEST_KeyNames: '[ci-${{ github.run_id }}-${{ github.run_attempt }}]'
```
[the remark this answers is not in the document]
Comment (sqin2019) [reply .github/workflows/ci-new.yml:47]: I can't quite tell why some of env vars with INTEG_TEST prefix, some of them dont, we need to find a way to be consistent.

[the remark this answers is not in the document]
Comment (sailorlqh) [reply .github/workflows/ci-new.yml]: I have removed the `INTEG_TEST` prefix, and also removed `INTEG_TEST_KEY_NAMES`, as it's not used.

File: test/integ-new/main_new_test.go (hunk 1 above)
Comment (sailorlqh) [reply test/integ-new/main_new_test.go]: Removed.

File: test/integ-new/main_new_test.go (hunk 3 above)
Comment (sailorlqh) [reply test/integ-new/main_new_test.go]: Combined.

File: test/integ-new/main_new_test.go (hunk 4 above)
Comment (sailorlqh) [reply test/integ-new/main_new_test.go]: I have moved the kms key clean up back to the workflow, so it won't affect other tests.

File: test/integ-new/main_new_test.go [hunk 6]
```diff
@@ -192,67 +203,101 @@ func TestAPIAndPublicKeyService(t *testing.T) {
...
-					t.Fatal(err)
-				}
-				t.Errorf("Got unexpected status code, got=%d want=%d, response=%s", got, want, string(b))
-			}
-		})
+	ctx := logging.WithLogger(context.Background(), logging.TestLogger(t))
+
+	uri := cfg.CertRotatorServiceAddr + "/"
+
+	kmsClient, err := kms.NewKeyManagementClient(ctx)
```
Comment (sqin2019) [inline test/integ-new/main_new_test.go:238]: Is kmsClient closed somewhere?

File: .github/workflows/ci-new.yml [hunk 7]
```diff
@@ -257,11 +268,25 @@ jobs:
       # Keys can not be deleted, we can destroy key versions to avoid billing.
       #
       # See: https://cloud.google.com/kms/docs/faq#cannot_delete.
-      - name: 'Clean up kms key'
-        if: '${{ always()  && steps.create_kms_key.outcome == ''success'' }}'
+      - name: 'Clean up kms key when integ test failed'
```
Comment (sqin2019) [inline .github/workflows/ci-new.yml:271]: For the cleanups, I wonder if we can do in code, as this is not scalable if we have more than 1 key versions.

File: test/integ-new/main_new_test.go [hunk 8]
```diff
@@ -192,67 +203,101 @@ func TestAPIAndPublicKeyService(t *testing.T) {
...
 
-			if got, want := resp.StatusCode, tc.wantStatusCode; got != want {
-				b, err := io.ReadAll(resp.Body)
-				if err != nil {
-					t.Fatal(err)
-				}
-				t.Errorf("Got unexpected status code, got=%d want=%d, response=%s", got, want, string(b))
-			}
-		})
+	ctx := logging.WithLogger(context.Background(), logging.TestLogger(t))
```
Comment (sqin2019) [inline test/integ-new/main_new_test.go:234]: Why not `ctx := context.Background() `, we have been using everyelses, does it provide any additional benefit?

**Comment by sailorlqh** [comment]
Updated to use a new method to do the key clean up. This new method can deleted all active keys after after all tests are finished. See example result below.

image

File: .github/workflows/ci-new.yml (hunk 7 above)
Comment (sailorlqh) [reply .github/workflows/ci-new.yml]: Please see below comment.

File: test/integ-new/main_new_test.go (hunk 6 above)
Comment (sailorlqh) [reply test/integ-new/main_new_test.go:237]: Done

File: .github/workflows/ci-new.yml [hunk 9]
```diff
@@ -268,25 +268,19 @@ jobs:
...
-              --key ${{ env.KMS_KEY_NAME }} \
-              --keyring ${{ env.KMS_KEYRING_ID }} \
-              --location global \
-              --project ${{ env.PROJECT_ID }}
+          KEY_VERSIONS=$(gcloud kms keys versions list \
+            --location=global \
+            --keyring=${{ env.KMS_KEYRING_ID }} \
+            --key=${{ env.KMS_KEY_NAME }} \
+            --project=${{ env.PROJECT_ID }} \
+            --filter="state=enabled" \
```
Comment (sqin2019) [inline .github/workflows/ci-new.yml:280]: 1. condition: always()  && steps.create_kms_key.outcome == ''success''
2. there are other states we care about too - https://cloud.google.com/kms/docs/key-states, why not change it to when state != destroy or destroy scheduled.

Comment (sailorlqh) [reply .github/workflows/ci-new.yml]: done

**Review by sqin2019 — APPROVED** [review]

**Pull request merged**